# The Monty Hall Problem

Here's a fun and perhaps surprising statistical riddle, and a good way to get some practice writing python functions
In a gameshow, contestants try to guess which of 3 closed doors contain a cash prize (goats are behind the other two doors). Of course, the odds of choosing the correct door are 1 in 3. As a twist, the host of the show occasionally opens a door after a contestant makes his or her choice. This door is always one of the two the contestant did not pick, and is also always one of the goat doors (note that it is always possible to do this, since there are two goat doors). At this point, the contestant has the option of keeping his or her original choice, or swtiching to the other unopened door. The question is: is there any benefit to switching doors? The answer surprises many people who haven't heard the question before.


We can answer the problem by running simulations in Python. We'll do it in several parts.
First, write a function called simulate_prizedoor. This function will simulate the location of the prize in many games -- see the detailed specification below:

In [1]:
import numpy as np

In [17]:
def simulate_prizedoor(nsim):
    return np.random.randint(0 ,3 ,(nsim))


Next, write a function that simulates the contestant's guesses for nsim simulations. Call this function simulate_guess. The specs:

In [19]:
def simulate_guess(nsim):
    return np.zeros(nsim, dtype=np.int)

write a win_percentage function that takes an array of guesses and prizedoors, and returns the percent of correct guesses

In [7]:
def win_percentage(guesses, prizedoors):
    return 100 * (guesses == prizedoors).mean()

In [26]:
nsim= 10000
#keep guesses
print "Win percentage when keeping original door"
print win_percentage(simulate_prizedoor(nsim), simulate_guess(nsim))

Win percentage when keeping original door
32.35


Next, write a function, goat_door, to simulate randomly revealing one of the goat doors that a contestant didn't pick.

In [27]:
def goat_door(prizedoors, guesses):
    
    result = np.random.randint(0, 3, prizedoors.size)
    while True:
        bad = (result == prizedoors) | (result == guesses)
        if not bad.any():
            return result
        result[bad] = np.random.randint(0, 3, bad.sum())


Write a function, switch_guess, that represents the strategy of always switching a guess after the goat door is opened.

In [28]:
def switch_guess(guesses, goatdoors):
    result = np.zeros(guesses.size)
    switch = {(0, 1): 2, (0, 2): 1, (1, 0): 2, (1, 2): 0, (2, 0): 1, (2, 1): 0}
    for i in [0, 1, 2]:
        for j in [0, 1, 2]:
            mask = (guesses == i) & (goatdoors == j)
            if not mask.any():
                continue
            result = np.where(mask, np.ones_like(result) * switch[(i, j)], result)
    return result

Compute the percentage after the contestant switches his door after a goat door is revealed. 

In [29]:
pd = simulate_prizedoor(nsim)
guess = simulate_guess(nsim)
goats = goat_door(pd, guess)
guess = switch_guess(guess, goats)
print "Win percentage when switching doors"
print win_percentage(pd, guess).mean()

Win percentage when switching doors
67.2


Hence
Win percentage when keeping original door
32.35

Win percentage when switching doors
67.2